certbot/tests/main_test.py

```diff
@@ -194,5 +211,833 @@ def test_failure(self, mock_make_or_verify):
+        mock_choose_accounts.return_value = self.accs[1]
+        self.assertEqual((self.accs[1], None), self._call())
+        self.assertEqual(
+            set(mock_choose_accounts.call_args[0][0]), set(self.accs))
+        self.assertEqual(self.accs[1].id, self.config.account)
+        self.assertTrue(self.config.email is None)
+
+    @mock.patch('certbot.client.display_ops.get_email')
+    def test_no_accounts_no_email(self, mock_get_email):
+        mock_get_email.return_value = 'foo@bar.baz'
+
+        with mock.patch('certbot.main.client') as client:
+            client.register.return_value = (
+                self.accs[0], mock.sentinel.acme)
+            self.assertEqual((self.accs[0], mock.sentinel.acme), self._call())
+        client.register.assert_called_once_with(
+            self.config, self.account_storage, tos_cb=mock.ANY)
+
+        self.assertEqual(self.accs[0].id, self.config.account)
+        self.assertEqual('foo@bar.baz', self.config.email)
+
+    def test_no_accounts_email(self):
+        self.config.email = 'other email'
+        with mock.patch('certbot.main.client') as client:
+            client.register.return_value = (self.accs[1], mock.sentinel.acme)
+            self._call()
+        self.assertEqual(self.accs[1].id, self.config.account)
+        self.assertEqual('other email', self.config.email)
+
+
+class DuplicativeCertsTest(storage_test.BaseRenewableCertTest):
+    """Test to avoid duplicate lineages."""
+
+    def setUp(self):
+        super(DuplicativeCertsTest, self).setUp()
+        self.config.write()
+        self._write_out_ex_kinds()
+
+    def tearDown(self):
+        shutil.rmtree(self.tempdir)
+
+    @mock.patch('certbot.util.make_or_verify_dir')
+    def test_find_duplicative_names(self, unused_makedir):
+        from certbot.main import _find_duplicative_certs
+        test_cert = test_util.load_vector('cert-san.pem')
+        with open(self.test_rc.cert, 'wb') as f:
+            f.write(test_cert)
+
+        # No overlap at all
+        result = _find_duplicative_certs(
+            self.cli_config, ['wow.net', 'hooray.org'])
+        self.assertEqual(result, (None, None))
+
+        # Totally identical
+        result = _find_duplicative_certs(
+            self.cli_config, ['example.com', 'www.example.com'])
+        self.assertTrue(result[0].configfile.filename.endswith('example.org.conf'))
+        self.assertEqual(result[1], None)
+
+        # Superset
+        result = _find_duplicative_certs(
+            self.cli_config, ['example.com', 'www.example.com', 'something.new'])
+        self.assertEqual(result[0], None)
+        self.assertTrue(result[1].configfile.filename.endswith('example.org.conf'))
+
+        # Partial overlap doesn't count
+        result = _find_duplicative_certs(
+            self.cli_config, ['example.com', 'something.new'])
+        self.assertEqual(result, (None, None))
+
+
+class MainTest(unittest.TestCase):  # pylint: disable=too-many-public-methods
```

Comment:
That sounds like a good strategy. But I think I will save breaking this into separate verb tests for another PR. This PR just gets the low hanging fruit, like the parsing tests.